client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java

```diff
@@ -712,6 +713,29 @@ public void resolveStorage(boolean objectStorage) {
       this.dataFormat.initialize(remoteCacheManager, name, isObjectStorage);
    }
 
+   @Override
+   public void resolveStorage(MediaType key, MediaType value, boolean objectStorage) {
+      // Set the storage first and initialize the current data format.
+      // We need this to check if the key type match.
+      resolveStorage(objectStorage);
```

Comment:
Not really. We need it before, so the client key type is retrieved below since it might fall back to the configured marshall registry. Ideally, it would initialize just this first time here.